src/main/java/meteordevelopment/meteorclient/renderer/text/Font.java

```diff
@@ -72,29 +54,83 @@ public Font(ByteBuffer buffer, int height) {
-                    packedChar.xoff2(),
-                    packedChar.yoff2(),
-                    packedChar.x0() * ipw,
-                    packedChar.y0() * iph,
-                    packedChar.x1() * ipw,
-                    packedChar.y1() * iph,
-                    packedChar.xadvance()
-                ));
-            }
+        // Preload basic ASCII characters
+        preloadAsciiCharacters();
+    }
+
+    private void preloadAsciiCharacters() {
+        STBTTPackedchar.Buffer cdata = STBTTPackedchar.create(128); // Basic Latin
+
+        // create the pack range
+        STBTTPackRange.Buffer packRange = STBTTPackRange.create(1);
+        packRange.put(STBTTPackRange.create().set(height, 32, null, 128, cdata, (byte) 2, (byte) 2));
+        packRange.flip();
+
+        // pack font ranges
+        STBTruetype.stbtt_PackFontRanges(packContext, buffer, 0, packRange);
+
+        // Load character data into charMap
+        for (int i = 0; i < cdata.capacity(); i++) {
+            STBTTPackedchar packedChar = cdata.get(i);
+
+            float ipw = 1f / size; // pixel width and height
+            float iph = 1f / size;
+
+            charMap.put(i + 32, new CharData(
+                packedChar.xoff(),
+                packedChar.yoff(),
+                packedChar.xoff2(),
+                packedChar.yoff2(),
+                packedChar.x0() * ipw,
+                packedChar.y0() * iph,
+                packedChar.x1() * ipw,
+                packedChar.y1() * iph,
+                packedChar.xadvance()
+            ));
         }
+
+        // Create texture
+        createTexture();
+    }
+
+    private void loadCharacter(int codePoint) {
+        if (packedChars.containsKey(codePoint)) return;
+
+        STBTTPackedchar.Buffer cdata = STBTTPackedchar.create(1);
+
+        // create the pack range
+        STBTTPackRange.Buffer packRange = STBTTPackRange.create(1);
+        packRange.put(STBTTPackRange.create().set(height, codePoint, null, 1, cdata, (byte) 2, (byte) 2));
+        packRange.flip();
+
+        // pack font ranges
+        STBTruetype.stbtt_PackFontRanges(packContext, buffer, 0, packRange);
+
+        STBTTPackedchar packedChar = cdata.get(0);
+
+        float ipw = 1f / size; // pixel width and height
+        float iph = 1f / size;
+
+        charMap.put(codePoint, new CharData(
+            packedChar.xoff(),
+            packedChar.yoff(),
+            packedChar.xoff2(),
+            packedChar.yoff2(),
+            packedChar.x0() * ipw,
+            packedChar.y0() * iph,
+            packedChar.x1() * ipw,
+            packedChar.y1() * iph,
+            packedChar.xadvance()
+        ));
+
+        packedChars.put(codePoint, packedChar);
+
+        // Re-create texture
+        createTexture();
+    }
+
+    private void createTexture() {
+        texture = new ByteTexture(size, size, bitmap, ByteTexture.Format.A, ByteTexture.Filter.Linear, ByteTexture.Filter.Linear);
```

Comment:
`size` and `bitmap` are constants, there's not really any reason to create a new object rather than modifying `ByteTexture` to make the `upload` method accessible